src/audio/dcblock/dcblock_hifi3.c

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <andrula.song@intel.com>
+
+#include <stdint.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/dcblock/dcblock.h>
+
+#ifdef DCBLOCK_HIFI3
+
+#include <xtensa/tie/xt_hifi3.h>
+LOG_MODULE_DECLARE(dcblock, CONFIG_SOF_LOG_LEVEL);
+
+static inline ae_int32x2  dcblock_cal(ae_int32x2 R, ae_int32x2 state_x, ae_int32x2 state_y,
+				      ae_int32x2 sample)
+{
+	ae_int64 out, temp;
+
+	/* R: Q2.30, y_prev: Q1.31 the result is Q2.62 */
+	temp = AE_MULF32S_LL(R, state_y);
+	out = AE_SUB64(AE_MOVAD32_L(sample), AE_MOVAD32_L(state_x));
+	/* shift out to 2.62 */
+	out = AE_ADD64S(AE_SLAI64S(out, 31), temp);
+	/* shift out to 1.63 */
+	return AE_ROUND32F64SSYM(AE_SLAI64S(out, 1));
+}
+
+/* Setup circular for component source */
+static inline void dcblock_set_circular(const struct audio_stream __sparse_cache *source)
+{
+	/* Set source as circular buffer 0 */
+	AE_SETCBEGIN0(source->addr);
+	AE_SETCEND0(source->end_addr);
+}
+
+#if CONFIG_FORMAT_S16LE
+static void dcblock_s16_default(const struct comp_dev *dev,
+				const struct audio_stream __sparse_cache *source,
+				const struct audio_stream __sparse_cache *sink,
+				uint32_t frames)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	ae_int16 *src = (ae_int16 *)source->r_ptr;
+	ae_int16 *dst = (ae_int16 *)sink->w_ptr;
+	ae_int16 *in;
+	ae_int16 *out;
+	ae_int32x2 R, state_x, state_y, sample;
+	ae_int16x4 in_sample, out_sample;
+	int ch, i, n;
+	int nch = source->channels;
+	const int inc = nch * sizeof(ae_int16);
+	int samples = nch * frames;
+
+	dcblock_set_circular(source);
+	while (samples) {
+		n = audio_stream_samples_without_wrap_s16(sink, dst);
+		n = MIN(n, samples);
+		for (ch = 0; ch < nch; ch++) {
+			in = src + ch;
+			out = dst + ch;
+			state_x = cd->state[ch].x_prev;
+			state_y = cd->state[ch].y_prev;
+			R = cd->R_coeffs[ch];
+			for (i = 0; i < n; i += nch) {
+				/* Load a 16 bit sample*/
+				AE_L16_XC(in_sample, in, inc);
+				/* store the 16 bit sample to high 16bit of 32bit register*/
+				sample = AE_CVT32X2F16_32(in_sample);
+				state_y = dcblock_cal(R, state_x, state_y, sample);
+				state_x = sample;
+				out_sample = AE_ROUND16X4F32SSYM(state_y, state_y);
+				AE_S16_0_XP(out_sample, out, inc);
+			}
+			cd->state[ch].x_prev = state_x;
+			cd->state[ch].y_prev = state_y;
+		}
+		samples -= n;
+		dst = audio_stream_wrap(sink, dst + n);
+		src = audio_stream_wrap(source, src + n);
```

Comment:
I think you could avoid this `src` pointer wrap in is circular modified by AE_L16_XC()?

Reply:
but since I processed signal by channel, so for each channel, I have to record the initial address for each channel

Reply:
Or do you more like to remove the usage of circular buffer for both src and dst?

Reply:
No, not removal. I was thinking a dummy XC load intrinsic to modify in -> src for next round with single instruction. Though the saving would be small since there's only one or zero sink wraps. The current code is cleaner so it's OK this way.